crates/bevy_pbr/src/atmosphere/functions.wgsl

```diff
@@ -0,0 +1,208 @@
+#define_import_path bevy_pbr::atmosphere::functions
+
+#import bevy_pbr::atmosphere::{
+    types::Atmosphere,
+    bindings::{
+        atmosphere, settings, view, lights, transmittance_lut, transmittance_lut_sampler, 
+        multiscattering_lut, multiscattering_lut_sampler, sky_view_lut, sky_view_lut_sampler,
+        aerial_view_lut, aerial_view_lut_sampler
+    },
+    bruneton_functions::{transmittance_lut_r_mu_to_uv, transmittance_lut_uv_to_r_mu, ray_intersects_ground},
+}
+
+// CONSTANTS
+
+const PI: f32 = 3.141592653589793238462;
+const TAU: f32 = 6.283185307179586476925;
+const FRAC_PI: f32 = 0.31830988618379067153; // 1 / π
+const FRAC_3_16_PI: f32 = 0.0596831036594607509; // 3 / (16π)
+const FRAC_4_PI: f32 = 0.07957747154594767; // 1 / (4π)
+
+// LUT UV PARAMATERIZATIONS
+
+fn multiscattering_lut_r_mu_to_uv(r: f32, mu: f32) -> vec2<f32> {
+    let u = 0.5 + 0.5 * mu;
+    let v = saturate((r - atmosphere.bottom_radius) / (atmosphere.top_radius - atmosphere.bottom_radius)); //TODO
+    return vec2(u, v);
+}
+
+fn multiscattering_lut_uv_to_r_mu(uv: vec2<f32>) -> vec2<f32> {
+    let r = mix(atmosphere.bottom_radius, atmosphere.top_radius, uv.y);
+    let mu = uv.x * 2 - 1;
+    return vec2(r, mu);
+}
+
+fn sky_view_lut_lat_long_to_uv(lat: f32, long: f32) -> vec2<f32> {
+    let u = long * FRAC_PI + 0.5;
+    let v = sqrt(2 * abs(lat) * FRAC_PI) * sign(lat) * 0.5 + 0.5;
+    return vec2(u, v);
+}
+
+fn sky_view_lut_uv_to_lat_long(uv: vec2<f32>) -> vec2<f32> {
+    let long = (uv.x - 0.5) * TAU;
+    let v_minus_half = uv.y - 0.5;
+    let lat = TAU * (v_minus_half * v_minus_half) * sign(v_minus_half);
+    return vec2(lat, long);
+}
+
+// LUT SAMPLING
+
+fn sample_transmittance_lut(r: f32, mu: f32) -> vec3<f32> {
+    let uv = transmittance_lut_r_mu_to_uv(r, mu);
+    return textureSampleLevel(transmittance_lut, transmittance_lut_sampler, uv, 0.0).rgb;
+}
+
+fn sample_multiscattering_lut(r: f32, mu: f32) -> vec3<f32> {
+    let uv = multiscattering_lut_r_mu_to_uv(r, mu);
+    return textureSampleLevel(multiscattering_lut, multiscattering_lut_sampler, uv, 0.0).rgb;
+}
+
+fn sample_sky_view_lut(ray_dir: vec3<f32>) -> vec3<f32> {
+    let lat_long = ray_dir_to_lat_long(ray_dir);
+    let uv = sky_view_lut_lat_long_to_uv(lat_long.x, lat_long.y);
+    return textureSampleLevel(sky_view_lut, sky_view_lut_sampler, uv, 0.0).rgb;
```

Comment:
Inverse trigonometric functions are *really* slow on GPU. Can you use a cubemap LUT instead? Directional data is basically why cubemap textures exist.

Reply:
I believe the author did it this way because the latitude is parameterized to put most of the detail near the horizon, which *does* have a noticeable effect on the final result. There's probably room to speed up my math here, but I'd want to make sure a cubemap could preserve that detail.

Reply:
I should check how Hillaire does this, since apparently it's good enough for unreal